### PR TITLE
Added support for broadcasting

### PR DESCRIPTION
**Pull request by shoatman**
- Defined platform independent interface for broadcasting
- Implemented AndroidBroadcaster

Tests to be added Tomorrow

### CODE REVIEW

File: common/src/main/java/com/microsoft/identity/common/AndroidPlatformComponents.java [hunk 1]
```diff
@@ -223,6 +226,15 @@ protected AndroidPlatformComponents(@NonNull final Context applicationContext,
         );
     }
 
+    @Override
+    public IBroadcaster getBroadcaster() throws ClientException {
```
Comment (shahzaibj) [inline common/src/main/java/com/microsoft/identity/common/AndroidPlatformComponents.java]: This method probably needs to be `synchronized`

File: common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import lombok.NonNull;
+
+@AllArgsConstructor
+public class AndroidBroadcaster implements IBroadcaster {
+
+    @NonNull
+    private Context mContext;
+
+    @Override
+    public void sendBroadcast(@NonNull String broadcastId, @Nullable PropertyBag propertyBag) {
```
Comment (shahzaibj) [inline common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java]:
```suggestion
    public void sendBroadcast(@NonNull final String broadcastId, @Nullable final PropertyBag propertyBag) {
```

File: common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        Intent intent = new Intent();
```
Comment (shahzaibj) [inline common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java]:
```suggestion
        final Intent intent = new Intent();
```

File: common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+import androidx.annotation.Nullable;
+
+import com.microsoft.identity.common.java.util.IBroadcaster;
+import com.microsoft.identity.common.java.util.ported.PropertyBag;
+
+import lombok.AllArgsConstructor;
+import lombok.NonNull;
+
+@AllArgsConstructor
+public class AndroidBroadcaster implements IBroadcaster {
```
Comment (shahzaibj) [inline common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java:41]: javadoc

File: common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+    @NonNull
+    private Context mContext;
+
+    @Override
+    public void sendBroadcast(@NonNull String broadcastId, @Nullable PropertyBag propertyBag) {
+        Intent intent = new Intent();
+        intent.setAction(broadcastId);
+        if(propertyBag != null) {
+            for (String key :
```
Comment (shahzaibj) [inline common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java]:
```suggestion
            for (final String key :
```

File: common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+                    propertyBag.keySet()) {
```
Comment (shahzaibj) [inline common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java]: The formatting looks a bit off. I think that code on second line can be on the same line

File: common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java [hunk 7]
```diff
@@ -0,0 +1,56 @@
...
+import com.microsoft.identity.common.java.util.ported.PropertyBag;
+
+import lombok.AllArgsConstructor;
+import lombok.NonNull;
+
+@AllArgsConstructor
+public class AndroidBroadcaster implements IBroadcaster {
+
+    @NonNull
+    private Context mContext;
```
Comment (shahzaibj) [inline common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java]:
```suggestion
    private final Context mContext;
```

File: common4j/src/main/com/microsoft/identity/common/java/util/IBroadcaster.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+package com.microsoft.identity.common.java.util;
+
+import com.microsoft.identity.common.java.util.ported.PropertyBag;
+
+/**
+ * An interface via which to access the ability to send device OS profile wide broadcasts
```
Comment (shahzaibj) [inline common4j/src/main/com/microsoft/identity/common/java/util/IBroadcaster.java]:
```suggestion
 * An interface via which to access the ability to send device OS profile wide broadcasts.
```

File: common4j/src/main/com/microsoft/identity/common/java/util/IBroadcaster.java [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+ * (These broadcasts are not currently restricted to a particular set of receivers)
```
Comment (shahzaibj) [inline common4j/src/main/com/microsoft/identity/common/java/util/IBroadcaster.java]:
```suggestion
 * (These broadcasts are not currently restricted to a particular set of receivers).
```

File: common4j/src/main/com/microsoft/identity/common/java/util/IBroadcaster.java [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+ */
+public interface IBroadcaster {
+
+    void sendBroadcast(String broadcastId, PropertyBag propertyBag);
```
Comment (shahzaibj) [inline common4j/src/main/com/microsoft/identity/common/java/util/IBroadcaster.java:38]: javadoc

File: common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java (hunk 6 above)
Comment (shoatman) [reply common/src/main/java/com/microsoft/identity/common/internal/platform/AndroidBroadcaster.java]: Interestingly... this was how the auto completion in Android Studio formatted it, but I tend to agree with you.

File: common/src/main/java/com/microsoft/identity/common/AndroidPlatformComponents.java (hunk 1 above)
Comment (shoatman) [reply common/src/main/java/com/microsoft/identity/common/AndroidPlatformComponents.java]: Good catch.  Thanks

**Review by shahzaibj — APPROVED** [review]

**Review by gitaumoses4 — APPROVED** [review]

**Pull request merged**